src/core_plugins/kibana/public/dashboard/panel/panel_utils.js

```diff
@@ -63,4 +66,18 @@ export class PanelUtils {
     }, 0);
     return ++maxId;
   }
+
+  static getPanelIndexPatterns(panels, objectLoadersForDashboard) {
+    return Promise.map(panels, (panel) => {
+      return loadSavedObject(objectLoadersForDashboard, panel)
+        .then(({ savedObj }) => {
+          if (savedObj.vis) {
+            return savedObj.vis.indexPattern;
+          } else if (savedObj.searchSource) {
+            return savedObj.searchSource.get('index');
+          }
+        });
+    })
+      .then(indexPatterns => _.uniq(indexPatterns));
+  }
```

Comment:
You could do something like the following to avoid reloading all the saved objects:

https://github.com/Stacey-Gammon/kibana/commit/bc6390f797287461a8b029bc7670a23ecc8b15a2